test/common/operation.go

```diff
@@ -0,0 +1,221 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/knative/eventing/test/base"
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	pkgTest "github.com/knative/pkg/test"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// LabelNamespace labels the given namespace with the labels map.
+func (client *Client) LabelNamespace(labels map[string]string) error {
+	namespace := client.Namespace
+	nsSpec, err := client.Kube.Kube.CoreV1().Namespaces().Get(namespace, metav1.GetOptions{})
+	if err != nil && errors.IsNotFound(err) {
+		return err
+	}
+	if nsSpec.Labels == nil {
+		nsSpec.Labels = map[string]string{}
+	}
+	for k, v := range labels {
+		nsSpec.Labels[k] = v
+	}
+	_, err = client.Kube.Kube.CoreV1().Namespaces().Update(nsSpec)
+	return err
+}
+
+// SendFakeEventToChannel will send the given event to the given channel.
+func (client *Client) SendFakeEventToChannel(senderName, channelName string, event *base.CloudEvent) error {
+	namespace := client.Namespace
+	channel, err := client.Eventing.EventingV1alpha1().Channels(namespace).Get(channelName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	return client.sendFakeEventToAddressable(senderName, channel.Status.Address, event)
+}
+
+// SendFakeEventToBroker will send the given event to the given broker.
+func (client *Client) SendFakeEventToBroker(senderName, brokerName string, event *base.CloudEvent) error {
+	namespace := client.Namespace
+	broker, err := client.Eventing.EventingV1alpha1().Brokers(namespace).Get(brokerName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	return client.sendFakeEventToAddressable(senderName, broker.Status.Address, event)
+}
+
+// sendFakeEventToAddressable will create a sender pod, which will send the given event to the given url.
+func (client *Client) sendFakeEventToAddressable(senderName string, addr duckv1alpha1.Addressable, event *base.CloudEvent) error {
```

Comment:
Channel and Broker should both be Addressable, so you should be able to feed the result from SendFakeEventToX directly to this as:

```
func (client *Client) sendFakeEventToAddressable(senderName, event *base.CloudEvent)
    func(duckv1alpha1.AddressableType, error) error {
  return func(obj duckv1alpha1.AddressableType, err error) {
    if err != nil {
      return fmt.Errorf("Failed to read target from apiserver: %v", err)
    }
    ...
  }
}
```

So then you can write:

```
sender := client.sendFakeEventToAddressable(senderName, event)
return sender(client.Eventing.EventingV1alpha1().Brokers(client.Namespace).Get(brokerName, metav1.GetOptions{}))
```

Reply:
Hmm, I got an error saying "cannot use broker `Broker` as type `AddressableType` in argument to send" if I make this change. And as we are not saving the updated reference for brokers and channels in the code, I don't see much benefit to use duck-type here. 
Please correct me if I'm wrong or misunderstood. Thanks.